R/oauth.R

```diff
@@ -0,0 +1,175 @@
+oauth2_server_app <- function(
+  app_name = "Third-Party App",
+  client_secret = "client_secret",
+  client_id = "client_id"
+  ) {
+  # Create app
+  app <- new_app()
+  app$locals$listener <- local_app_process(useless())
+  # Register third-party app
+  app$locals$apps <- data.frame(
+    app_name = app_name,
+    client_secret = client_secret,
+    client_id = client_id
+  )
+
+  # First step, asking the user for access
+  app$get("/authorize", function(req, res) {
+
+    # Missing information either client secret or client ID
+    if (is.null(req$query$client_secret) || is.null(req$query$client_id)) {
+      res$
+        # How would I set a status message, maybe in the body
+        send_status(400L)
+    } else {
+
+      corresponding_app <- res$app$locals$apps[
+        res$app$locals$apps$client_secret == req$query$client_secret &&
+          res$app$local$apps$client_id == req$query$client_id,]
+
+      # Wrong client ID or token
+      if (nrow(corresponding_app) == 0) {
+        res$
+          # How would I set a status message
+          send_status(400L)
+      } else {
+
+        code <- sodium::bin2hex(sodium::random(15))
+
+        req$app$locals$codes <- c(req$app$locals$codes, code)
+
+        app <- corresponding_app$app_name[1]
+        state <- req$query$state
+        url <- req$query$redirect_uri
+
+        listener_url <- req$app$locals$listener$url()
+
+        txt <- glue::glue(
+        "<body><p>
+        Hey would you authorize the Third-Party App {app} to access your account?
+        </p>
+        <p><a href='{listener_url}/kill'>Yes, allow!</a>
+</p></body>")
+
+        file <- tempfile()
+        writeLines(
+          txt,
+          file)
+        browseURL(file)
+
+       accepted <- unlist(httr::content(httr::GET(httr::modify_url(listener_url, path = "/killed"))))
+       while(!accepted) {
+         Sys.sleep(0.1)
+         accepted <- unlist(httr::content(httr::GET(httr::modify_url(listener_url, path = "/killed"))))
+       }
+
+        res$
+          set_header("location", glue::glue("{url}/cb?state={state}&code={code}"))$
+          send_status(302L)
+
+      }
+    }
+  })
+
+    app$post("/token", function(req, res) {
+
+      if (req$query$code %in% req$app$locals$codes) {
+
+        token <- sodium::bin2hex(sodium::random(5))
+
+        req$app$locals$tokens <- c(req$app$locals$tokens, token)
+        req$app$locals$codes <- req$app$locals$codes[req$app$locals$codes != req$query$code]
+
+        res$
+          send_json(list(access_token = token, expires_in = 3600L),
```

Comment:
note that there is no code ensuring it does expire, yet.

Reply:
Since we don't do anything with the issues tokens, I don't think we need to fix this.